experimental/ixdtf/src/parser.rs

```diff
@@ -0,0 +1,400 @@
+}
+
+impl DateTimeParser {
+
+    fn empty() -> Self {
+        DateTimeParser {
+            year : None,
+            month : None,
+            day : None,
+            hour : None,
+            minute : None,
+            second : None,
+            nano_second : None
+        }
+    }
+
+    fn parse_date_year(idx :usize, s: &str) -> (usize, Option<i32>) {
+        match &s[idx..(idx + 1)] {
+            "+" | "-" => {
+                match s[idx..(idx + 7)].parse::<i32>() {
+                    Ok(year) => ((idx + 7), Some(year)),
+                    Err(_) => (idx, None)
+                }
+            }
+            _ => match s[idx..(idx + 4)].parse::<i32>() {
+                Ok(year) => ((idx + 4), Some(year)),
+                Err(_) => (idx, None)
+            }
+        }
+    }
+
+    fn parse_date_month(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(month) => {
+                if month > 0 && month <= 12 {
+                    ((idx + 2), Some(month))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_date_day(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(day) => {
+                if day > 0 && day <= 31 {
+                    ((idx + 2), Some(day))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_hour(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(hour) => {
+                if hour <= 24 {
+                    ((idx + 2), Some(hour))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_minute(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(minute) => {
+                if minute < 60 {
+                    ((idx + 2), Some(minute))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_second(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(second) => {
+                if second <= 60 {
+                    ((idx + 2), Some(second))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_fractional_part(idx :usize, s: &str) -> (usize, Option<i32>) {
+        match s[idx..].parse::<i32>() {
+            Ok(fraction) => {
+                (s.len(), Some(fraction))
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+
+    pub fn parse(s: String) -> Self {
```

Comment:
Nit: Take the borrowed type `&str` as an argument unless you need to take ownership (which you don't).

Suggestion: Take `&[u8]` instead of `String` or `&str` since the string is always ASCII.